### ISSUE TITLE
Missing support for threads

### ISSUE THREAD

**Issue by ptersilie** [ykjit/yk#794]
We currently can't compiled interpreters that use threads, e.g. via `pthread_create`. The reason for this is the shadowstack: we use the shadowstack to share variables between AOT and JIT. The shadow stack needs to be a thread local, so that we don't get races between multiple threads executing a trace.

In order to fix this we need to clone the shadowstack for each new thread that is created and update the pointer living in thread local `shadow_stack_0`. We can do this fairly easily by finding calls to `pthread_create` in the interpreter and inject code that clones and updates the shadow stack. However, this won't work for threads created within libraries whose IR we can't see. We briefly talked about adding a wrapper around `pthread_create` or similar, but I don't remember the details.

**Comment by ltratt** [comment]
> In order to fix this we need to clone the shadowstack for each new thread

Not sure we need to *clone* the shadowstack so much as create an entirely new one?

**Comment by ltratt** [comment]
@Pavel-Durov Interested in having a look at this?

**Comment by Pavel-Durov** [comment]
Yes, I am interested :)
Looking into it!

**Comment by ltratt** [comment]
@Pavel-Durov Excellent :) There are at least two possibilities:

1. Statically detect calls to `pthread_create`.
2. Intercept calls to `pthread_create` at run-time.

I think the first of those is fundamentally difficult and, probably, impossible in the general case. The second case might be doable in various ways (`LD_PRELOAD`-esque tricks and others). @ptersilie and @vext01 might have further ideas, but I suspect none of us know the best way of doing this. [Since intercepting thread creation is a fairly obvious thing to want to do, there might even be a "nice" way of doing this, but I don't know what that might be!]

**Comment by vext01** [comment]
'ld --wrap' may help as long as the interpreter doesn't already use that
facility.

**Comment by ltratt** [comment]
> 'ld --wrap' may help as long as the interpreter doesn't already use that
facility.

I hadn't thought of this! We could probably do this quite easily by adding it to the flags in https://github.com/ykjit/yk/blob/master/bin/yk-config#L56.

**Comment by Pavel-Durov** [comment]
Am I understanding correctly:

`pthread_create` is a function defined in `ykllvm` 

If we add `ld --wrap pthread_create` to [yk-config](https://github.com/ykjit/yk/blob/master/bin/yk-config#L56) we can make sure that all the `yk` integrated `pthread_create` calls will call `wrap_pthread_create` instead - and that's how we intercept
?

**Comment by ltratt** [comment]
`pthread_create` is a POSIX function for creating a new thread (try `man pthread_create`). The problem with creating a new thread is that we also need to create a new shadow stack (currently we don't do that). 

> If we add ld --wrap pthread_create to [yk-config](https://github.com/ykjit/yk/blob/master/bin/yk-config#L56) we can make sure that all the yk integrated pthread_create calls will call wrap_pthread_create instead - and that's how we intercept

Roughly yes, though it's slightly more subtle than that. Here's my man page for `ld --wrap`:

```
     --wrap=symbol
             Redirect symbol references to __wrap_symbol and __real_symbol
             references to symbol.
```

In other words, it's less "wrapping" then "renaming".

**Comment by jacob-hughes** [comment]
You might find it helps to see how Boehm does this to intercept threads for GC. I couldn’t find much documentation on this, but this file might be a good place to start:

https://github.com/ivmai/bdwgc/blob/master/include/gc/gc_pthread_redirects.h

**Comment by ltratt** [comment]
> Where would be a good place to put it in yk project?

I don't know if Boehm's approach is something we can adopt or whether we have to do something else to intercept thread creation and create a shadow stack, so the answer to your question is unclear until you tell us which approach is best :) In an ideal world we'd transparently do things in a way that interpreter authors don't even need to know about. In a less-ideal-but-still-OK world we'd force some linker flags on the interpreter author. In a not-very-ideal world we'd force them to make manual changes.

**Comment by Pavel-Durov** [comment]
We decided to go with `ld` wrapper function for `pthread_create`.
Here's my thread interception poc - https://github.com/Pavel-Durov/thread-intercept.poc/tree/main/ld-wrap.
YK location for thread interception will be `ykrt`

**Comment by Pavel-Durov** [comment]
I am testing the `pthread_create` interception with yklua
And I got this error:
```
ld.lld: error: Unable to add shadow stack: No support for threads yet!
```

coming from [ShadowStack.cpp:197](https://github.com/ykjit/ykllvm/blob/d3caa6bab38c7f182591f1f9bbf40ffa7ff1d3b6/llvm/lib/Transforms/Yk/ShadowStack.cpp#L197)

So it looks like we know how to detect `pthread_create` :)

Is the idea here with `pthread_create` interception to detect the call before it gets to ykllvm [ShadowStack.cpp:197](https://github.com/ykjit/ykllvm/blob/d3caa6bab38c7f182591f1f9bbf40ffa7ff1d3b6/llvm/lib/Transforms/Yk/ShadowStack.cpp#L197)

**Comment by ltratt** [comment]
> So it looks like we know how to detect pthread_create :)

Yes, but only for code compiled with ykllvm -- so external C code, for example, won't hit this check.

**Comment by Pavel-Durov** [comment]
ok, got it
